File: app/server/app/routers/expenses.py
"""
Router para operaciones CRUD de Expenses
"""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List, Optional
import uuid
from datetime import datetime

from app.database import get_db
from app.models import Expense as ExpenseModel
from app.schemas import Expense, ExpenseCreate

router = APIRouter()


@router.post("/", response_model=Expense)
def create_expense(
    expense: ExpenseCreate,
    db: Session = Depends(get_db)
):
    """Crear un nuevo gasto"""
    db_expense = ExpenseModel(
        id=str(uuid.uuid4()),
        category=expense.category,
        item=expense.item,
        amount=expense.amount,
        description=expense.description,
        attachment_url=expense.attachment_url,
        store_id=expense.store_id,
        register_id=expense.register_id,
        date=expense.date or datetime.now().strftime("%Y-%m-%d"),
        user_id=expense.user_id,
        created_at=datetime.now()
    )
    db.add(db_expense)
    db.commit()
    db.refresh(db_expense)
    return db_expense


@router.get("/", response_model=List[Expense])
def get_expenses(
    skip: int = 0,
    limit: int = 100,
    store_id: Optional[str] = None,
    category: Optional[str] = None,
    date: Optional[str] = None,
    db: Session = Depends(get_db)
):
    """Obtener todos los gastos"""
    query = db.query(ExpenseModel)
    
    if store_id:
        query = query.filter(ExpenseModel.store_id == store_id)
    if category:
        query = query.filter(ExpenseModel.category == category)
    if date:
        query = query.filter(ExpenseModel.date == date)
    
    return query.order_by(ExpenseModel.created_at.desc()).offset(skip).limit(limit).all()


@router.get("/{expense_id}", response_model=Expense)
def get_expense(expense_id: str, db: Session = Depends(get_db)):
    """Obtener un gasto por ID"""
    expense = db.query(ExpenseModel).filter(
        ExpenseModel.id == expense_id
    ).first()
    
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    return expense


@router.put("/{expense_id}", response_model=Expense)
def update_expense(
    expense_id: str,
    expense: ExpenseCreate,
    db: Session = Depends(get_db)
):
    """Actualizar un gasto"""
    db_expense = db.query(ExpenseModel).filter(
        ExpenseModel.id == expense_id
    ).first()
    
    if not db_expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    for key, value in expense.dict().items():
        setattr(db_expense, key, value)
    
    db.commit()
    db.refresh(db_expense)
    return db_expense


@router.delete("/{expense_id}")
def delete_expense(expense_id: str, db: Session = Depends(get_db)):
    """Eliminar un gasto"""
    expense = db.query(ExpenseModel).filter(
        ExpenseModel.id == expense_id
    ).first()
    
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    db.delete(expense)
    db.commit()
    return {"message": "Expense deleted successfully"}


@router.get("/stats/by-category", response_model=dict)
def get_expenses_by_category(
    store_id: Optional[str] = None,
    date_from: Optional[str] = None,
    date_to: Optional[str] = None,
    db: Session = Depends(get_db)
):
    """Obtener estadísticas de gastos agrupados por categoría"""
    query = db.query(ExpenseModel)
    
    if store_id:
        query = query.filter(ExpenseModel.store_id == store_id)
    if date_from:
        query = query.filter(ExpenseModel.date >= date_from)
    if date_to:
        query = query.filter(ExpenseModel.date <= date_to)
    
    expenses = query.all()
    
    stats = {}
    for expense in expenses:
        category = expense.category
        if category not in stats:
            stats[category] = {"count": 0, "total": 0.0}
        stats[category]["count"] += 1
        stats[category]["total"] += expense.amount
    
    return stats

File: app/server/src/routes/expenses.ts
import { Router, Request, Response } from 'express';
import { v4 as uuidv4 } from 'uuid';
import { Op } from 'sequelize';
import { Expense } from '../models/Expense';
import { expenseCreateSchema } from '../schemas/expense';

const router = Router();

// Crear un nuevo gasto
router.post('/', async (req: Request, res: Response) => {
  try {
    const validatedData = expenseCreateSchema.parse(req.body);
    
    const expense = await Expense.create({
      id: uuidv4(),
      ...validatedData,
      date: validatedData.date || new Date().toISOString().split('T')[0],
      created_at: new Date(),
    });

    res.status(201).json(expense);
  } catch (error: any) {
    if (error.name === 'ZodError') {
      res.status(400).json({ error: 'Datos inválidos', details: error.errors });
    } else {
      res.status(500).json({ error: error.message });
    }
  }
});

// Obtener todos los gastos
router.get('/', async (req: Request, res: Response) => {
  try {
    const { skip = 0, limit = 100, store_id, category, date } = req.query;
    
    const where: any = {};
    if (store_id) where.store_id = store_id;
    if (category) where.category = category;
    if (date) where.date = date;

    const expenses = await Expense.findAll({
      where,
      limit: Number(limit),
      offset: Number(skip),
      order: [['created_at', 'DESC']],
    });

    res.json(expenses);
  } catch (error: any) {
    res.status(500).json({ error: error.message });
  }
});

// Obtener un gasto por ID
router.get('/:expense_id', async (req: Request, res: Response) => {
  try {
    const expense = await Expense.findByPk(req.params.expense_id);
    
    if (!expense) {
      return res.status(404).json({ error: 'Expense not found' });
    }

    res.json(expense);
  } catch (error: any) {
    res.status(500).json({ error: error.message });
  }
});

// Actualizar un gasto
router.put('/:expense_id', async (req: Request, res: Response) => {
  try {
    const validatedData = expenseCreateSchema.parse(req.body);
    const expense = await Expense.findByPk(req.params.expense_id);
    
    if (!expense) {
      return res.status(404).json({ error: 'Expense not found' });
    }

    await expense.update(validatedData);
    res.json(expense);
  } catch (error: any) {
    if (error.name === 'ZodError') {
      res.status(400).json({ error: 'Datos inválidos', details: error.errors });
    } else {
      res.status(500).json({ error: error.message });
    }
  }
});

// Eliminar un gasto
router.delete('/:expense_id', async (req: Request, res: Response) => {
  try {
    const expense = await Expense.findByPk(req.params.expense_id);
    
    if (!expense) {
      return res.status(404).json({ error: 'Expense not found' });
    }

    await expense.destroy();
    res.json({ message: 'Expense deleted successfully' });
  } catch (error: any) {
    res.status(500).json({ error: error.message });
  }
});

// Obtener estadísticas de gastos por categoría
router.get('/stats/by-category', async (req: Request, res: Response) => {
  try {
    const { store_id, date_from, date_to } = req.query;
    
    const where: any = {};
    if (store_id) where.store_id = store_id;
    if (date_from && date_to) {
      where.date = {
        [Op.between]: [date_from, date_to],
      };
    } else if (date_from) {
      where.date = {
        [Op.gte]: date_from,
      };
    } else if (date_to) {
      where.date = {
        [Op.lte]: date_to,
      };
    }

    const expenses = await Expense.findAll({ where });
    
    const stats: Record<string, { count: number; total: number }> = {};
    expenses.forEach((expense) => {
      const category = expense.category;
      if (!stats[category]) {
        stats[category] = { count: 0, total: 0 };
      }
      stats[category].count += 1;
      stats[category].total += expense.amount;
    });

    res.json(stats);
  } catch (error: any) {
    res.status(500).json({ error: error.message });
  }
});

export default router;

